src/core/dense_set.cc

```diff
@@ -0,0 +1,558 @@
+ * covered the range [000-111] (all keys in that case).
+ * Returns: next cursor or 0 if reached the end of scan.
+ * cursor = 0 - initiates a new scan.
+ *
+ * Therefore for any given Scan cursor x which was run when capacity_log_ was m, if the
+ * next Scan is run when capacity_log_ is n the current bucket to check is x << (n - m)
+ *
+ * When kFastScan is enabled this optimization is used.
+ * To be able to make this comparison the cursor must store the capacity_log_ from when
+ * Scan is run. To do this the (capacity_log_)th bit is set in the cursor, this will always
+ * be the highest bit set since the number of buckets is 2**(capacity_log_) and therefore
+ * the greatest index is (2 ** capacity_log_) - 1.
+ *
+ * When using this optimization the maximum number of buckets is 2 ** 31 instead of 2 ** 32
+ */
+
+// Extract the previous capacity_log_ from the cursor
+static inline std::pair<size_t, uint32_t> cursor_to_idx(uint32_t cursor) {
+  DCHECK(cursor != 0);
+  size_t expected_capacity_log = 0;
+  uint32_t expected_capacity_log_bitmask = 1;
+
+  while ((cursor >> expected_capacity_log) != 1) {
+    ++expected_capacity_log;
+    expected_capacity_log_bitmask <<= 1;
+  }
+
+  return {expected_capacity_log, expected_capacity_log_bitmask};
+}
+
+uint32_t DenseSet::Scan(uint32_t cursor, const ItemCb& cb) const {
+  uint32_t entries_idx = 0;
+
+  // if the fast scan optimization is enabled
+  if constexpr (kFastScan) {
+    DCHECK(entries_.size() <= (1ULL << 31));
+    // not a new scan
+    if (cursor != 0) {
+      // extract the previous capacity_log_ and actual bucket index from the cursor
+      auto [cursor_capacity_log, cursor_bitmask] = cursor_to_idx(cursor);
+      entries_idx = cursor & ~cursor_bitmask;
+
+      // Since DenseSet does not currently support shrinking the current log2(capacity)
+      // should never be less than the log2(capacity) of the cursor from a previous Scan run
+      DCHECK(cursor_capacity_log <= capacity_log_);
+
+      // if the capacity has grown significantly use the above described optmization
+      // to skip as many buckets as possible
+      if (cursor_capacity_log < capacity_log_) {
+        entries_idx <<= capacity_log_ - cursor_capacity_log;
+      }
+    }
+  }
+
+  // Shrinkage is not currently supported by the DenseSet type
+  DCHECK(entries_idx < 1ULL << capacity_log_);
+
+  // skip empty entries
+  while (entries_idx < entries_.size() && entries_[entries_idx].IsEmpty()) {
+    ++entries_idx;
+  }
+
+  // when scanning add all entries in a given chain
+  for (ConstChainIterator it = ChainConstBegin(entries_idx); it != ChainConstEnd(entries_idx);
+       ++it) {
+    cb(it->GetObject());
+  }
+
+  // move to the next index for the next scan and check if we are done
+  ++entries_idx;
+  if (entries_idx >= entries_.size()) {
+    return 0;
+  }
+
+  // encode the current capacity_log_ in the cursor if needed
+  if constexpr (kFastScan) {
+    entries_idx |= 1 << capacity_log_;
```

Comment:
that I fail to understand, how is storing the old capacity_log_ helps you

Reply:
originally it was needed to find the change in a capacity_log between scan calls to be able to determine if the current bucket iterator could "jump" as described in the previous comment. this wasn't clean so I re-implemented it by shifting using `(32-capacity_log_)` to hopefully make it easier to understand

Reply:
look at my previous comment. The alg should work without keeping the previous shift value.